---
title: "scRNASeq_Mafe"
author: "Marcelo Hurtado"
date: "2023-04-12"
output: html_document
---

Tutorial: https://satijalab.org/seurat/articles/pbmc3k_tutorial.html

Libraries
```{r setup, include=FALSE}
library(Seurat)
library(SeuratDisk)
library(tidyr)
library(dplyr)
library(org.Hs.eg.db)
library(ggplot2)
library(patchwork)
```

Setup the Seurat Object
```{r}
Convert("NK_cells.h5ad", dest = "h5seurat", overwrite = T)
seurat_anndata = LoadH5Seurat("NK_cells.h5seurat",  assays = "RNA")
```

```{r}
x = levels(seurat_anndata@meta.data$sample)
ColumnData_Vanderbilt <- read.csv("~/LungPredict/RawFiles/ColumnData_Vanderbilt.csv")
ColumnData_Vanderbilt = ColumnData_Vanderbilt[which(ColumnData_Vanderbilt$pt_ID %in% x),]
```

```{r}
# Dimensionality reduction
pbmc <- NormalizeData(seurat_anndata, normalization.method = "LogNormalize", scale.factor = 10000)
cll.combined <- RunPCA(pbmc, features = VariableFeatures(object = pbmc), verbose = FALSE)
print(cll.combined[["pca"]], dims = 1:5, nfeatures = 5)
DimHeatmap(cll.combined, dims = 1, cells = 500, balanced = T)
DimHeatmap(cll.combined, dims = 2, cells = 500, balanced = T)
DimHeatmap(cll.combined, dims = 1:2, cells = 500, balanced = T)
# Explain dimensionality of the data
ElbowPlot(cll.combined) #Principal components ranked by the percentage of the variance explained

# Clustering
cll.combined <- FindNeighbors(cll.combined, dims = 1:10) #based on the previous elbow plot

# Defining the granularity of the clusters
cll.combined <- FindClusters(cll.combined, resolution = c(0.05, 0.1, 0.3, 0.5, 0.8), verbose = FALSE)
#View(cll.combined@meta.data)
DimPlot(cll.combined, group.by = "RNA_snn_res.0.05", label = T)
DimPlot(cll.combined, group.by = "RNA_snn_res.0.1", label = T) # Resolution above 0.1 shows overlapping
DimPlot(cll.combined, group.by = "RNA_snn_res.0.3", label = T)
DimPlot(cll.combined, group.by = "RNA_snn_res.0.5", label = T)
DimPlot(cll.combined, group.by = "RNA_snn_res.0.8", label = T)
# Defining identity of clusters
Idents(cll.combined) # Default resolution
Idents(cll.combined) <- "RNA_snn_res.0.1" # Resolution sets at 0.1
Idents(cll.combined) # 6 clusters

# Non-linear dimensionality reduction
cll.combined <- RunUMAP(cll.combined, dims = 1:10) #increase the dimension if necessary
# Grouped UMAP
DimPlot(cll.combined, reduction = "umap")
# Splited by day UMAP
DimPlot(cll.combined, reduction = "umap", split.by = "orig.ident", label = TRUE)

```

QC and selecting cells for further analysis
```{r}
#We calculate mitochondrial QC metrics with the PercentageFeatureSet() function, which calculates the percentage of counts originating from a set of features
seurat_anndata[["percent.mt"]] <- PercentageFeatureSet(seurat_anndata, pattern = "^MT-")
VlnPlot(seurat_anndata, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.
plot1 <- FeatureScatter(seurat_anndata, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(seurat_anndata, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```

Normalizing the data
```{r}
pbmc <- NormalizeData(seurat_anndata, normalization.method = "LogNormalize", scale.factor = 10000)
```

Identification of highly variable features
```{r}
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(pbmc), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(pbmc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2
```

Scaling the data
```{r}
#The results of this are stored in pbmc[["RNA"]]@scale.data
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)
#Therefore, the default in ScaleData() is only to perform scaling on the previously identified variable features (2,000 by default). To do this, omit the features argument in the previous function call, i.e.
```
Perform linear dimensional reduction
```{r}
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
# Examine and visualize PCA results a few different ways
print(pbmc[["pca"]], dims = 1:5, nfeatures = 5)
VizDimLoadings(pbmc, dims = 1:2, reduction = "pca")
DimPlot(pbmc, reduction = "pca")
DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)
DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)
```
Determine the dimensionality of the dataset
```{r}
# NOTE: This process can take a long time for big datasets, comment out for expediency. More
# approximate techniques such as those implemented in ElbowPlot() can be used to reduce
# computation time
pbmc <- JackStraw(pbmc, num.replicate = 100)
pbmc <- ScoreJackStraw(pbmc, dims = 1:20)
JackStrawPlot(pbmc, dims = 1:15)
ElbowPlot(pbmc)
```

Cluster the cells
```{r}
pbmc = seurat_anndata
pbmc <- FindNeighbors(pbmc, dims = 1:10)
pbmc <- FindClusters(pbmc, resolution = 0.5)
# Look at cluster IDs of the first 5 cells
head(Idents(pbmc), 5)
```
Non linear dimensional reduction (UMAP/tSNE)
```{r}
pbmc <- RunUMAP(pbmc)
DimPlot(pbmc, reduction = "umap")
```
Save processed file
```{r}
saveRDS(pbmc, file = "../Mafe_scRNAseq.rds")
```

Find differentially expressed features (cluster biomarkers)
```{r}
# find all markers of cluster 2
cluster2.markers <- FindMarkers(pbmc, ident.1 = 2, min.pct = 0.25)
head(cluster2.markers, n = 5)
# find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers <- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3), min.pct = 0.25)
head(cluster5.markers, n = 5)
# find markers for every cluster compared to all remaining cells, report only the positive ones
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
pbmc.markers = pbmc.markers %>%
  group_by(cluster) %>%
  slice_max(n = 2, order_by = avg_log2FC)

#Plotting DE genes
VlnPlot(pbmc, features = c("LINC01857", "NR4A2"))
# you can plot raw counts as well
VlnPlot(pbmc, features = c("LINC01857", "NR4A2"), slot = "counts", log = TRUE)
FeaturePlot(pbmc, features = c("LINC01857", "NR4A2", "PTPN6", "SCIMP", "LMNA", "TUBA1A", "IGKV3D-20", "IGKV3D-15","RPS4Y1", "FKBP4", "IGLC3", "IGLC2", "	IGKV1D-39", "IGKV1-39", "IL32", "CD3E", "DUSP4", "ZEB2", "IGHG3", "XIST"))
pbmc.markers %>%
    group_by(cluster) %>%
    top_n(n = 10, wt = avg_log2FC) -> top10
DoHeatmap(pbmc, features = top10$gene) + NoLegend()
```

```{r}
FeaturePlot(pbmc, features = c("NKG7", "GNLY"))
```

Cell type identity to clusters
```{r}
new.cluster.ids <- c("Naive CD4 T", "CD14+ Mono", "Memory CD4 T", "B", "CD8 T", "FCGR3A+ Mono", "NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```

Save session
```{r}
saveRDS(pbmc, file = "../output/pbmc3k_final.rds")
```


