---
title: "scRNASeq_Mafe"
author: "Marcelo Hurtado"
date: "2023-04-12"
output: html_document
---

```{r setup, include=FALSE}
library(Seurat)
library(SeuratDisk)
library(tidyr)
library(dplyr)
library(ggplot2)
library(decoupleR)
library(OmnipathR)
library(SingleR)
library(pheatmap)
```

```{r}
#Convert("all_samples.h5ad", dest = "h5seurat", overwrite = T)
seurat_anndata = LoadH5Seurat("all_samples.h5seurat",  assays = "RNA")

levels(seurat_anndata@meta.data$sample)
clinical.data <- data.frame(read.csv("/home/marcelo/LungPredict1/RawFiles/ColumnData_Vanderbilt.csv", row.names = 1))
clinical.data <- clinical.data[,c(1,5,6,9,11,13)]

coldata = clinical.data[clinical.data$pt_ID%in%levels(seurat_anndata@meta.data$sample),]
#14958 14965 11817 13634 15467 12929  8356 12889 15002
#"R3388_YZ_1"  "R3388_YZ_2"  "R3388_YZ_10" "R3388_YZ_11" "R3388_YZ_18" "R3388_YZ_28" "R3388_YZ_56" "R4163_YZ_16" "R4163_YZ_25"
```

Clustering
```{r}
DimPlot(seurat_anndata, reduction="umap")
ElbowPlot(seurat_anndata) #determine dimensionality of data
seurat_anndata = FindNeighbors(seurat_anndata, dims = 1:20)
seurat_anndata = FindClusters(seurat_anndata, resolution = 1)
DimPlot(seurat_anndata, reduction="umap", label = T)
```

Manual annotation 

Find markers (one cluster vs all others)
```{r}
markers = FindAllMarkers(seurat_anndata,
               logfc.threshold = 0.25, #min logfc 
               min.pct = 0.1, #genes that are detected on 50% frequency across clusters
               only.pos = T,
               test.use = 'wilcox',
               slot = 'counts')
head(markers)
```

```{r, fig.width = 14, fig.height = 11, fig.align = 'center'}
features = c('NKG7', 'IL7R', 'HLA-DRA','IGHG1','BATF','IGFBP7', 'TYROBP', 'RAMP2','TYMS','TPSAB1','WFDC2','SFRP21')
RidgePlot(seurat_anndata, features = features, ncol = 3)
VlnPlot(seurat_anndata, features = features)
FeaturePlot(seurat_anndata, features = features)
```

Automatic cell annotation (ATLAS)
```{r}
ref = celldex::HumanPrimaryCellAtlasData()
single_counts = GetAssayData(seurat_anndata, slot='counts')
pred = SingleR(test = single_counts, 
        ref = ref,
        labels = ref$label.main)
head(pred)
seurat_anndata$singleR.labels = pred$labels[match(rownames(seurat_anndata@meta.data), rownames(pred))]
DimPlot(seurat_anndata, reduction = 'umap', group.by = 'singleR.labels')
head(pred$scores)
plotScoreHeatmap(pred)
plotDeltaDistribution(pred)
tab = table(Assigned = pred$labels, Clusters = seurat_anndata$seurat_clusters)
pheatmap(log10(tab+10), color=colorRampPalette(c('white', 'blue'))(10))
```
Extract NK cluster (cluster = 8) for further analysis
```{r}
Idents(seurat_anndata) <- "seurat_clusters" 
NK = subset(seurat_anndata, idents = "8", invert = FALSE)
DimPlot(NK, reduction="umap")
```
NK clustering
```{r}
ElbowPlot(NK) #determine dimensionality of data
NK<- FindNeighbors(NK, dims = 1:10)
NK<- FindClusters(NK, algorithm= 1, resolution = 0.15, verbose = T, graph.name = "RNA_snn")
NK<- RunUMAP(NK, dims = 1:10)
DimPlot(NK, reduction="umap", label = T)
```

Differential expressed markers across clusters
```{r}
markers = FindMarkers(NK, ident.1 = "0", ident.2 = "1")
head(markers)
features = c('GZMB', 'FCGR3A', 'PRF1','IL7R','GNLY','CTSW')
#features = c('CCL3', 'TIGIT', 'CD69','IRF8','IL2RB','CD160') markers from Pierre-Paul
FeaturePlot(NK, features = features)
VlnPlot(NK, features = features)
RidgePlot(NK, features = features, ncol = 2)
saveRDS(NK, "nk.rds") 
```