---
title: "Single cell analysis"
author: "Marcelo Hurtado"
date: "2023-07-10"
output: html_document
---

Load packages and functions
```{r}
source(paste0(getwd(),"/environment_set.R"))
libraries_set()
```

To install SeuratDisk, firt
- sudo apt-get install libhdf5-dev

Load seurat object 
```{r}
#Convert("all_samples.h5ad", dest = "h5seurat", overwrite = T)
seurat_anndata = LoadH5Seurat("all_samples.h5seurat",  assays = "RNA")
```

Check samples that have both single cell and bulk data
```{r}
levels(seurat_anndata@meta.data$sample)
clinical.data <- data.frame(read.csv("/home/marcelo/LungPredict1/RawFiles/ColumnData_Vanderbilt.csv", row.names = 1))
clinical.data <- clinical.data[,c(1,5,6,9,11,13)]

coldata = clinical.data[clinical.data$pt_ID%in%levels(seurat_anndata@meta.data$sample),]
#14958 14965 11817 13634 15467 12929  8356 12889 15002
#"R3388_YZ_1"  "R3388_YZ_2"  "R3388_YZ_10" "R3388_YZ_11" "R3388_YZ_18" "R3388_YZ_28" "R3388_YZ_56" "R4163_YZ_16" "R4163_YZ_25"
```

Cluster identification
```{r}
DimPlot(seurat_anndata, reduction="umap")
ElbowPlot(seurat_anndata) #determine dimensionality of data
seurat_anndata = FindNeighbors(seurat_anndata, dims = 1:20)
seurat_anndata = FindClusters(seurat_anndata, resolution = 1)
DimPlot(seurat_anndata, reduction="umap", label = T)
```

Cell annotation (ATLAS)
```{r}
ref = celldex::HumanPrimaryCellAtlasData()
single_counts = GetAssayData(seurat_anndata, slot='counts')
pred = SingleR(test = single_counts, 
        ref = ref,
        labels = ref$label.main)
seurat_anndata$singleR.labels = pred$labels[match(rownames(seurat_anndata@meta.data), rownames(pred))]
DimPlot(seurat_anndata, reduction = 'umap', group.by = 'singleR.labels')
head(pred$scores)
plotScoreHeatmap(pred)
plotDeltaDistribution(pred)
tab = table(Assigned = pred$labels, Clusters = seurat_anndata$seurat_clusters)
pheatmap(log10(tab+10), color=colorRampPalette(c('white', 'blue'))(10))
Idents(seurat_anndata) = seurat_anndata@meta.data$singleR.labels #Rename clusters
```

Extract NK cluster for further analysis
```{r}
DimPlot(NK, reduction="umap")
ElbowPlot(NK) #determine dimensionality of data
NK<- FindNeighbors(NK, dims = 1:20)
NK<- FindClusters(NK, algorithm= 1, resolution = 0.15, verbose = T, graph.name = "RNA_snn")
NK<- RunUMAP(NK, dims = 1:20)
DimPlot(NK, reduction="umap", label = T, pt.size = 0.5) + 
  NoLegend() + ggtitle('NK cells')
```

Extract macrophages cluster for further analysis
```{r}
macrophages = subset(seurat_anndata, idents = 12, invert = FALSE)
DimPlot(macrophages, reduction="umap")
ElbowPlot(macrophages) #determine dimensionality of data
macrophages<- FindNeighbors(macrophages, dims = 1:10)
macrophages<- FindClusters(macrophages, algorithm= 1, resolution = 0.1, verbose = T, graph.name = "RNA_snn")
macrophages<- RunUMAP(macrophages, dims = 1:10)
DimPlot(macrophages, reduction="umap", label = T, pt.size = 0.5) + 
  NoLegend() + ggtitle('Macrophages cells')
```

Classify subclusters NK and macrophages
```{r}
Idents(NK) 
new.cluster.ids <- c("Cluster1", "Cluster2", "Cluster3")
names(new.cluster.ids) <- levels(NK)
NK <- RenameIdents(NK, new.cluster.ids)

Idents(macrophages) 
new.cluster.ids <- c("Cluster1", "Cluster2", "Cluster3")
names(new.cluster.ids) <- levels(macrophages)
macrophages <- RenameIdents(macrophages, new.cluster.ids)
```

Find all markers across the different NK subclusters
```{r}
markers = FindAllMarkers(NK,
               logfc.threshold = 0.25, #min logfc 
               min.pct = 0.1, #genes that are detected on 50% frequency across clusters
               only.pos = T,
               test.use = 'wilcox',
               slot = 'counts')
head(markers[which(markers$cluster==0),], n =5)
head(markers[which(markers$cluster==1),], n =5)
head(markers[which(markers$cluster==2),], n =5)
head(markers[which(markers$cluster==3),], n =5)
features = c("NKG7", "IL7R", "TPSAB1", "UBE2C")
RidgePlot(NK, features = features, ncol = 2)
VlnPlot(NK, features = features, ncol = 2)
FeaturePlot(NK, features = features, ncol = 2)
```

Differential expressed markers across NK clusters 0 and 1
```{r}
markers = FindMarkers(NK, ident.1 = "0", ident.2 = "1")
head(markers, n=20)
features = c("NKG7", "GNLY", "KLRD1", "GZMH", "GZMB", "GZMA")
#features = c('CCL3', 'TIGIT', 'CD69','IRF8','IL2RB','CD160') markers from Pierre-Paul
FeaturePlot(NK, features = features, ncol = 3)
VlnPlot(NK, features = features, ncol = 3)
RidgePlot(NK, features = features, ncol = 3)
```

########################################################################

TFs inference from single cell data 

Extract TFs network from Dorothea database
```{r}
dorothea2viper_regulons <- function(df) {
  regulon_list <- split(df, df$tf)
  viper_regulons <- lapply(regulon_list, function(regulon) {
    tfmode <- stats::setNames(regulon$mor, regulon$target)
    list(tfmode = tfmode, likelihood = rep(1, length(tfmode)))
  })
    
  return(viper_regulons)
}
  
data("dorothea_hs", package = "dorothea")
regulons <- dorothea_hs %>%
  filter(confidence %in% c("A", "B", "C"))
```

Extracting only variable features from NK cluster to calculate TFs activity
```{r}
NK <- FindVariableFeatures(NK, selection.method = "vst")
var_feat = VariableFeatures(NK)
nk_top <- GetAssayData(NK)[var_feat,]

macrophages <- FindVariableFeatures(macrophages, selection.method = "vst")
var_feat = VariableFeatures(macrophages)
macrophages_top <- GetAssayData(macrophages)[var_feat,]
```

Run Univariate linear model (ULM) of TFs in NK object
```{r}
# Extract the normalized log-transformed counts
mat <- as.matrix(nk_top)
# Run ulm
acts <- run_ulm(mat=mat, net=regulons, .source='tf', .target='target',
                .mor='mor', minsize = 5)
head(acts)

# Extract the normalized log-transformed counts
mat <- as.matrix(macrophages_top)
# Run ulm
acts <- run_ulm(mat=mat, net=regulons, .source='tf', .target='target',
                .mor='mor', minsize = 5)
head(acts)

```

Save TFs results in Seurat object
```{r}
# Extract ulm and store it in tfsulm in pbmc
NK[['tfsulm']] <- acts %>%
  pivot_wider(id_cols = 'source', names_from = 'condition',
              values_from = 'score') %>%
  column_to_rownames('source') %>%
  Seurat::CreateAssayObject(.)

# Change assay
DefaultAssay(object = NK) <- "tfsulm"

# Scale the data
NK <- ScaleData(NK)
NK@assays$tfsulm@data <- NK@assays$tfsulm@scale.data
```

Clustering using protein activity
```{r}
NK<- FindNeighbors(NK, dims = 1:10)
NK<- FindClusters(NK, algorithm= 1, resolution = 0.15, verbose = T, graph.name = "RNA_snn")
NK<- RunUMAP(NK, dims = 1:10)
DimPlot(NK, reduction="umap", label = T, pt.size = 0.5) + 
  NoLegend() + ggtitle('NK cells')
```

Find all TFs markers of NK subclusters
```{r}
markers = FindAllMarkers(NK,
               logfc.threshold = 0.25, #min logfc 
               min.pct = 0.1, #genes that are detected on 50% frequency across clusters
               only.pos = T,
               test.use = 'wilcox',
               slot = 'counts')

head(markers[which(markers$cluster==0),], n =5)
head(markers[which(markers$cluster==1),], n =5)
head(markers[which(markers$cluster==2),], n =5)
head(markers[which(markers$cluster==3),], n =5)
features = c("STAT4", "SOX11", "FOSL2", "E2F4")

RidgePlot(NK, features = features, ncol = 2)
VlnPlot(NK, features = features, ncol = 2)
(FeaturePlot(NK, features = features, ncol = 2)& 
  scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red'))
```

Differential TFs across cluster 0 vs 1 NK
```{r}
markers = FindMarkers(NK, ident.1 = "0", ident.2 = "1")
head(markers, n=30)
features = c("STAT4", "TBX21", "FOSL2", "NR2F2", "BATF", "RUNX1","IRF4","BCL6")
(FeaturePlot(NK, features = features, ncol = 2)& 
  scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red'))
VlnPlot(NK, features = features, ncol = 2)
RidgePlot(NK, features = features, ncol = 2)
```

Visualization and comparison of results from gene expression and protein activity
```{r}
p1 <- DimPlot(NK, reduction = "umap", label = TRUE, pt.size = 0.5) + 
  NoLegend() + ggtitle('Cell types')
p2 <- (FeaturePlot(NK, features = c("STAT4")) & 
  scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')) +
  ggtitle('STAT4 activity')
DefaultAssay(object = NK) <- "RNA"
NK@assays$RNA@data <- NK@assays$RNA@scale.data
p3 <- (FeaturePlot(NK, features = c("STAT4")) & 
  scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')) + ggtitle('STAT4 expression')
DefaultAssay(object = NK) <- "tfsulm"
p1 | p2 | p3
```

Features plot of TFs found in bulk analysis
```{r}
(FeaturePlot(NK, features = c("TBX21","PAX5", "LYL1", "HIF1A", "CXXC4")) & 
  scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red'))
```

Save ouput file
```{r}
saveRDS(NK, "nk.rds")
```


